### ISSUE TITLE
Allow conditionial parameter input

### ISSUE THREAD

**Issue by gaelcolas** [PowerShell/Plaster#21]
It might be useful to allow the parameters to be requested based on conditions, similar to what you do in the content section.

That way, you could have context specific parameters such as:

```
if ! PLASTER_PARAM_useDHCP
- ask for vm_ip
- ask for subnet mask.
```

**Comment by rkeithhill** [comment]
Good idea!

**Comment by zloeber** [comment]
Submitted pull request to add this enhancement ([#255](https://github.com/PowerShell/Plaster/pull/255)

**Comment by michaeltlombardi** [comment]
It looks [like you determined](https://github.com/PowerShell/Plaster/pull/255#issuecomment-299555117) this functionality conflicts with the dynamic parameters implementation, @rkeithhill? Just wanted to cross-link that back to here if the issue is going to be closed for that reason.

**Comment by zloeber** [comment]
Technically dynamic parameters would be fundamentally incompatible with conditional plaster parameters. If conditional parameters were left in as dynamic parameters then the responsibility for dealing with that situation would be on the template author. Either this should be made clear in the documentation or plaster parameters with a conditional attributes can simply be left out of the dynamic parameter creation process. This will force  the interactive prompt for input  at invocation (assuming the condition is met).

Either way I question the requirement for dynamic parameters in this project. It is simply one form of interactivity that the project already provides (with its own prompts). It also doesn't allow for automatic platyps documentation generation of the function at build time. Perhaps it would make more sense to allow passing a hashtable of parameters instead.

**Comment by KevinMarquette** [comment]
I just ran into the need for this feature on a new template that I was creating and was surprised it didn't exist. It just felt to natural to add a condition to a parameter like we do for the content. 

My scenario is giving the user a choice to include something. If they choose to include it, prompt for additional information. 

After reading this thread and the pull request, I would fully expect as the template author, that I have to account for that. 

On that note, we could call it a `PromptCondition` to be more self documenting. Or process the conditional logic for each parameter in order and `$null` the values that fail the condition.

I will work past this by creating two different templates that duplicate a lot of content or just cut features that I planned on adding.

**Comment by rkeithhill** [comment]
@michaeltlombardi  I don't want to close this yet.  I'd like to have this capability as well. It provides "wizard-like" alternate paths of prompts.  I just didn't want to knee jerk implement something (which I almost did) without considering the impact on dynamic parameters.

I wonder if we could add an attribute to `<parameter>` along the lines of "dependsOnParameter" e.g.:

```xml
<parameters>
  <parameter name="AskOnlyOnLinux" 
             condition="$IsLinux" ... />
  <parameter name="Editor" type="choice" ... />
  <parameter name="VSCodeOptions" 
             condition="$PLASTER_PARAM_Editor -eq 'VSCode'" 
             dependsOnParameters="Editor"
             type="multichoice" ... />
</parameter>
```

If a conditional parameter depends on multiple parameter values, then those parameter names would be listed as comma separated: `dependsOnParameters="Editor, AskOnlyOnLinux"`.  In the Windows (or macOS) case, the AskOnlyOnLinux parameter would never get a value. In that case, the value of $PLASTER_PARAM_AskOnlyOnLinux would be `$null`.

We would start out providing dynamic parameters on manifest parameters that aren't conditional and those that are conditional (and condition evals to true) but do no "depend on" the user-supplied value of any other parameter.  Then, as parameter values are supplied by the user, we can evaluate the conditional parameters that "depend on" the value of other parameters and if those values have been provided, then add them to the dynamic parameter list (if their condition evals to true).  I think that just might work.  Need to test it out though.

Another option, and probably a better one from a template authors POV, is to just have Plaster find the dependent parameters by extracting the `PLASTER_PARAM_` variable references in the condition.  Then we don't have to worry about folks forgetting to provide a `dependsOnParameters` attribute.  Yeah, I think I like this better:

```xml
<parameters>
  <parameter name="AskOnlyOnLinux" 
             condition="$IsLinux" ... />
  <parameter name="Editor" type="choice" ... />
  <parameter name="VSCodeOptions" 
             condition="$PLASTER_PARAM_Editor -eq 'VSCode'" 
             type="multichoice" ... />
</parameter>
```

Thoughts?

**Comment by KevinMarquette** [comment]
My scenario was having a multiple choice selection. Then use that as a condition on future parameters.

Here is the exact parameter that I was trying to implement. 
 
    <parameter name='TemplateType' 
               type='choice' 
               default='2' 
               store='text' 
               prompt='Select the template type'>
      <choice label='&amp;Single TemplateFile' 
              help="Creates a template that contains a single TemplateFile" 
              value="Single"/>
      <choice label='&amp;Import from Directory' 
              help="Creates a template that will deploy a directory full of files" 
              value="Directory"/>
      <choice label='&amp;Empty' 
              help="Create an empty template with a manifest" 
              value="Empty"/>
    </parameter>
    <parameter name="FileName" 
               type="text" 
               prompt="Name of TemplateFile to create" 
               condition="$PLASTER_PARAM_TemplateType -eq 'Single'" />
    <parameter name="SourceFolder" 
               type="text" 
               prompt="Source folder path to build template from"
                condition="$PLASTER_PARAM_TemplateType -eq 'Directory'" />

**Comment by rkeithhill** [comment]
So in that scenario, tab completion would initially only list `-TemplateType`.  After that parameter value is specified, either `-FileName` or `-SourceFolder` would be available via tab completion depending on the value specified for `-TemplateType`.

**Comment by LaurentDardenne** [comment]
I wonder if the original need is not another concept than Parameter :
```xml
<parameters>
  <parameter name='Editor'
             type='choice'
             prompt='Select one of the supported script editors for better editor integration (or None):'
             default='0'
             store='text' >
      <choice label='&amp;None'
              help="No editor specified."
              value="None"/>
      <choice label='Visual Studio &amp;Code'
              help="Your editor is Visual Studio Code."
              value="VSCode"/>
  </parameter>
  
 <Condition condition="$PLASTER_PARAM_Editor -eq 'VSCode'" />
   <parameters>             
     <parameter name="VSCodeOptions" 
              type="multichoice" ... />
   </parameter>
   <!-- Nested Condition  ?  -->
  </parameter>              
</parameter>
```

**Comment by rkeithhill** [comment]
I don't think so.  I think this fits nicely into the current parameter concept.  It is just conditional. Like I said before, if it weren't for not having a good dynamic parameter solution (for automation use), I would merged in my previous attempt to do this.

**Comment by zloeber** [comment]
Is it even possible to have multiple dynamic parameters dependent upon one another with conditional logic? I recall trying to do this for some other pet project and running into issues. Or maybe I just wasn't trying hard enough :)

**Comment by rkeithhill** [comment]
Well, I thought it was because `RuntimeDefinedParameter` has both `IsSet` and `Value` properties.  But apparently these do not get set while dynamicparam is still being processed.  So, the idea of being able to "dynamically light up" dynamic parameters depending on the value of previously specified dynamic parameters is an apparent `no go`.   :-(

I guess we're back to having dynamicparam list **every parameter** even if certain parameters wind up not being used.  This can make `automating` the invocation of such templates a bit weird, requiring a few passes to determine which parameters to specify so you don't wind up getting prompts for unspecified parameters.

**Comment by zloeber** [comment]
Would it make sense to define all the parameters and maybe shove the conditional logic into the validatescript attribute (after stripping out the plaster specific prefixes of course)?

**Comment by rkeithhill** [comment]
Addressed with PR #255. Thanks @zloeber!
